### ISSUE TITLE
&& false and || true not being optimized

### ISSUE THREAD

**Issue by ncannasse** [HaxeFoundation/haxe#8278]
The following code:

```haxe
class Main {

	static function bar(v) {
		if( v == null && false ) throw "assert";
		if( v == null || true ) throw "OK";
	}
	
	static function main() {
		bar("");
	}
	
}
```

Outputs the following JS:

```js
Main.bar = function(v) {
	if(v == null && false) {
		throw new js__$Boot_HaxeError("assert");
	}
	if(v == null || true) {
		throw new js__$Boot_HaxeError("OK");
	}
};
```

**Comment by Simn** [comment]
This is already optimized with `-D analyzer-optimize`. Isn't that good enough?

**Comment by ncannasse** [comment]
I reported because this gives warnings with goggle closure. I also expect Haxe to optimize these with having to add extra compilation flags that might eventually trigger some other errors or change the code in more extensive manner (although I might be wrong here as I haven't played much with analyzer-optimize)

**Comment by markknol** [comment]
Wouldn't it be nice to just make `-D analyzer-optimize` default? I use it in all of my projects and never had issues with it.

**Comment by Simn** [comment]
Yeah I'm also not a fan of adding the same optimization in multiple places. This isn't even one that we could do completely locally because you have to consider the side-effectness of the other operand(s) and possibly convert the expression into an unconditional block.

**Comment by kevinresol** [comment]
> Wouldn't it be nice to just make `-D analyzer-optimize` default? I use it in all of my projects and never had issues with it.

In fact I think it is expected for Haxe 4, as it is mentioned in [the blog](https://haxe.org/blog/haxe-3.4-release/)

> To have extra optimizations the `-D analyzer-optimize` can be used ... in Haxe 4.0 it will be enabled by default.

**Comment by Simn** [comment]
Closing for reasons previously stated. If you want something optimized, enable the optimizer.

**Comment by ncannasse** [comment]
I beg to differ on this one. Because of conditional compilation this simple optimization was present in previous Haxe versions without any special compiler flag required. It's quite simple and does not require much work so I think it should be there by default instead of requiring some extra compiler flag.

**Comment by RealyUniqueName** [comment]
I don't know why anyone would want to not use `-D analyzer-optimize`. It's stable, it's fast, it's optimizing :)
Also `maybeSideEffect() || true` is not just a subject to eliminate. 
But `true || maybeSideEffect()` is. And it indeed looks easy to implement. 
Though, I'd prefer to not hack it just for the sake of `if(#if disableThis false && #end someCheck) {`. Instead one can puth whole `if` expression into `#if !disableThis`

**Comment by ncannasse** [comment]
Usually it's more   static inline var DEBUG = #if (debug || someDefine)
true #else false #end;
if( DEBUG || x > 0 ) doSomething();
